### PR TITLE
Support for basic sei payload parsing

### PR DESCRIPTION
**Pull request by jhnbwrs**
As well as more specific structure around the user_data_registered_itu_t_t35 payload

### CODE REVIEW

**Comment by jhnbwrs** [comment]
Give this a look, maybe provide any feedback.

File: src/h265_sei_parser.cc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  uint32_t payload_type = 0;
+  uint32_t ff_byte = 0xff;
+  
+  while(ff_byte == 0xff) {
+    bit_buffer->ReadBits(8, ff_byte); // f(8)
+    payload_type += ff_byte;
+  }
+
+  std::unique_ptr<SeiState> sei = nullptr;
+  if(static_cast<SeiType>(payload_type) == SeiType::user_data_registered_itu_t_t35) {
```
[the remark this answers is not in the document]
Comment (chemag) [reply src/h265_sei_parser.cc:44]: Also, for future foolproof, suggest:

```
switch (static_cast<SeiType>(payload_type)) {
    case SeiType::user_data_registered_itu_t_t35:
        ...

File: src/h265_sei_parser.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  } else {
+    sei = std::make_unique<SeiStateNotImplemented>();
+  }
+  sei->payload_type = static_cast<SeiType>(payload_type);
+
+  ff_byte = 0xff;
+  while(ff_byte == 0xff) {
+    bit_buffer->ReadBits(8, ff_byte); // f(8)
+    sei->payload_size += ff_byte;
+  }
```
Comment (chemag) [inline src/h265_sei_parser.cc:55]: I'd move this block to just line 43, just after the payload_type calculation. Both payload_size and payload_type should be treated the same. I'd suggest making them variables of the SeiState class (right now you only have payload_type)

File: test/h265_sei_parser_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+class H265SeiParserTest : public ::testing::Test {
+ public:
+  H265SeiParserTest() {}
+  ~H265SeiParserTest() override {}
+};
+
+TEST_F(H265SeiParserTest, TestUserDataRegisteredItuTT35Sei) {
+  // fuzzer::conv: data
+  const uint8_t buffer[] = {
+    0x04, 0x47, 0xB5, 0x00, 0x31, 0x47, 0x41, 0x39, 0x34, 0x03, 0x54, 0x00,
```
Comment (chemag) [inline test/h265_sei_parser_unittest.cc:25]: we're using non-caps hex (i.e. "0xb5" instead of "0xB5")

File: test/h265_sei_parser_unittest.cc [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+class H265SeiParserTest : public ::testing::Test {
+ public:
+  H265SeiParserTest() {}
+  ~H265SeiParserTest() override {}
+};
+
+TEST_F(H265SeiParserTest, TestUserDataRegisteredItuTT35Sei) {
+  // fuzzer::conv: data
+  const uint8_t buffer[] = {
```
Comment (chemag) [inline test/h265_sei_parser_unittest.cc:24]: curious where you got this SEI

File: include/h265_sei_parser.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    sei_manifest = 200,
+    sei_prefix_indication = 201,
+    annotated_regions = 202,
+    shutter_interval_info = 205,
+    not_set = 999
+};
+
+// A class for parsing out a supplemental enhancement information (SEI) data from
+// an H265 NALU.
+class H265SeiParser {
```
Comment (chemag) [inline include/h265_sei_parser.h:88]: I'd suggest 2+N classes here:

class #1:
```
H265SeiMessageParser {
   ...
   struct SeiMessageState {
...
```

class #2:
```
H265SeiPayloadParser {
    ...
     struct SeiPayloadState {
...
```

class #N:
```
H265SeiUserDataRegisteredItuTT35Parser {
    ...
     struct H265SeiUserDataRegisteredItuTT35State {
...
```

File: src/h265_sei_parser.cc [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+  // now read the payload:
+  sei->payload.resize(sei->payload_size);
+  for (size_t i = 0; i < sei->payload_size; ++i) {
+    bit_buffer->ReadUInt8(sei->payload[i]);
+  }
+
```
Comment (chemag) [inline src/h265_sei_parser.cc:62]: this is not how the standard defines SEI. 

sei_message() reads payload_type and payload_size. Then it sends the 2x parameters to the next class (sei_payload()).

sei_payload() just disptaches to the different classes (e.g. user_data_registered_itu_t_t35).

the different classes themselves read the payload. For example, user_data_registered_itu_t_t35 reads 2x values (itu_t_t35_country_code and itu_t_t35_country_code_extension_byte) and then the payload itself (itu_t_t35_payload_byte in this case).

File: src/h265_sei_parser.cc [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+const uint8_t* H265SeiParser::SeiStateUserData::get_userdata_payload(size_t& payload_size) const {
+  payload_size = payload.size() - userdata_payload_start_index;
+  return &(payload[userdata_payload_start_index]);
+}
+
+void H265SeiParser::SeiStateNotImplemented::parse_payload() {
+  // No-op
+  return;
```
Comment (chemag) [inline src/h265_sei_parser.cc:88]: I'm happy implementing a "unknown" sei_payload virtual parser that just puts all the bytes (per payload_size) in a single vector. But it has to read the payload here.

**Review by chemag — COMMENTED** [review]
some minor changes, otherwise lgtm

File: test/h265_sei_parser_unittest.cc (hunk 4 above)
Comment (jhnbwrs) [reply test/h265_sei_parser_unittest.cc:24]: Eventually I just found an h.264 sample that has some SEIs and attempted a software transcode with ffmpeg and libx265.  These User Data seis came along for the ride.  I inspected the h.264 and h.265 specs and there are a few seis that have the exact same structure between the codecs, user data is one of these.  Which may be one of the reasons it gets carried along on the transcode operation.  Most of the other seis that have exactly the same structure, I haven't yet been able to find on an h.264 sample and try to perform the same trick.  The list of seis that should be the same according to my research are as follows: filler_payload(3), 
user_data_registered_itu_t_t35(4), 
user_data_unregistered(5),  
progressive_refinement_segment_start(16),
progressive_refinement_segment_end(17), 
mastering_display_colour_volume(137), 
content_light_level_info(144), 
alternative_transfer_characteristics(147), 
ambient_viewing_environment(148), 
content_colour_volume(149), 
equirectangular_projection(150), 
cubemap_projection(151), 
sphere_rotation(154), 
omni_viewport(156), 
annotated_regions(202)

File: include/h265_sei_parser.h [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    depth_representation_info = 177,
+    multiview_scene_info = 178,
+    multiview_acquisition_info = 179,
+    multiview_view_position = 180,
+    alternative_depth_info = 181,
+    sei_manifest = 200,
+    sei_prefix_indication = 201,
+    annotated_regions = 202,
+    shutter_interval_info = 205,
+    not_set = 999
```
[the remark this answers is not in the document]
Comment (jhnbwrs) [reply include/h265_sei_parser.h]: Done

File: src/h265_sei_parser.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhnbwrs) [reply src/h265_sei_parser.cc]: Done and done

File: test/h265_sei_parser_unittest.cc (hunk 3 above)
Comment (jhnbwrs) [reply test/h265_sei_parser_unittest.cc]: Ok

File: include/h265_sei_parser.h (hunk 5 above)
Comment (jhnbwrs) [reply include/h265_sei_parser.h]: I think I captured what you were asking for here.  Go ahead and give it look over again and see what you think.

File: src/h265_sei_parser.cc (hunk 7 above)
Comment (jhnbwrs) [reply src/h265_sei_parser.cc]: Easy enough.

File: src/h265_sei_parser.cc (hunk 6 above)
Comment (jhnbwrs) [reply src/h265_sei_parser.cc]: Good points.  Hopefully I have resolved it to your satisfaction.

File: src/h265_sei_parser.cc (hunk 2 above)
Comment (jhnbwrs) [reply src/h265_sei_parser.cc]: Done

**Comment by jhnbwrs** [comment]
Tool output looks like this now when I parse an h265 stream and it comes upon one of the user data payloads:nal_unit { nal_unit_header { forbidden_zero_bit: 0 nal_unit_type: 39 nuh_layer_id: 0 nuh_temporal_id_plus1: 1 } nal_unit_payload { sei message { payload_type: 4 payload_size: 71 user_data_registered_itu_t_t35 sei { country_code: 181 country_code_extension: 0 payload_size: 70  } } } }

File: src/h265_sei_parser.cc [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+std::unique_ptr<H265SeiParser::SeiState> H265SeiParser::ParseSei(
+    rtc::BitBuffer* bit_buffer) noexcept {
+  // H265 SEI NAL Unit (access_unit_delimiter_rbsp()) parser.
+  // Section 7.3.5 ("Supplemental enhancement information message syntax") of the H.265
+  // standard for a complete description.
+
+  uint32_t payload_type = 0;
+  uint32_t ff_byte = 0xff;
+  
+  while(ff_byte == 0xff) {
```
[the remark this answers is not in the document]
Comment (jhnbwrs) [reply src/h265_sei_parser.cc]: Done -- I suspect the clang format fixed up all of these anyway.

**Comment by jhnbwrs** [comment]
actually not having the full SEI payload in the sei message is going to cause me some grief.  I have to reconstruct the sei message state payload bytes after parsing.... So either I need to add a requirement to implement a to_byte() function for each state, or I need to save the original byte stream before parsing so I can return that if I need to get the bytes version of the various sei_states.  So I probably have one more change to make after applying all of your suggestions.

**Comment by chemag** [comment]
can you please (a) merge the 2x patches and (b) rename the resulting patch using the traditional linux naming mechanism?

```
h265nal: <sentence>
```

Where <sentence> is something that will make grammatical sense in "if applied, this patch will <sentence>"

**Comment by jhnbwrs** [comment]
Ok, can do, would you mind if I also added serialize() function to the sei states so they can serialize themselves back out to byte arrays?

**Comment by chemag** [comment]
> Ok, can do, would you mind if I also added serialize() function to the sei states so they can serialize themselves back out to byte arrays?

Happy to hear your proposal, though let's do it in a separate patch, so we can merge the parsing support.

**Comment by chemag** [comment]
> actually not having the full SEI payload in the sei message is going to cause me some grief. I have to reconstruct the sei message state payload bytes after parsing.... So either I need to add a requirement to implement a to_byte() function for each state, or I need to save the original byte stream before parsing so I can return that if I need to get the bytes version of the various sei_states. So I probably have one more change to make after applying all of your suggestions.

I'd suggest to first finish the SEI parser (just the ITU T-35 and the "Unknown" used as default is fine). Then, let's discuss how to do message dumping (converting from protobuf structures back to binary streams). The latter needs to be designed carefully. Also, in order to be useful we probably need to add dump support for all the current supported sub-parsers.

**Comment by jhnbwrs** [comment]
Ok, sounds good.  I am going to close this PR now and open another one with the once sentence description you were asking for.  I won't include anything other than what is already in here.

**Pull request closed without merging**